### PR TITLE
Add ST_IsValid and geometry_invalid_reason functions

### PR DESCRIPTION
**Pull request by mbasmanova**
Add functions to check whether a given geometry is well formed and understand why if it is not.

- ST_IsValid function returns true iff input geometry value is well formed
- geometry_invalid_reason returns the reason why geometry is not well formed or null if it is

ST_IsValid function follows SQL/MM Part 3 standard. The standard doesn't have a function to return a reason for why geometry is not well formed, hence, geometry_invalid_reason name follows Presto convention.

https://postgis.net/docs/using_postgis_dbmanagement.html#OGC_Validity provides a good explanation of what it means for a geometry to be well formed.

The implementation is based on functionality provided by the Java Topology Suite (JTS). ESRI library is missing this functionality.

Existing geospatial functions encode geometry objects into Slices using ESRI shapefile format described in https://www.esri.com/library/whitepapers/pdfs/shapefile.pdf. This PR includes a new deserialization method that converts geometry Slice into JTS geometry object.

### CODE REVIEW

File: presto-docs/src/main/sphinx/functions/geospatial.rst [hunk 1]
```diff
@@ -159,6 +159,12 @@ Accessors
 
     Returns ``true`` if and only if the line is closed and simple.
 
+.. function:: ST_IsValid(Geometry) -> boolean
+
+    Returns ``true`` if and only if the input geometry is well formed.
+    geometry_is_valid_reason function returns the reason why geometry is
```
Comment (findepi) [inline presto-docs/src/main/sphinx/functions/geospatial.rst]: `s/geometry_is_valid_reason/geometry_is_*in*valid_reason/` or `geometry_invalid_reason`?

Comment (mbasmanova) [reply presto-docs/src/main/sphinx/functions/geospatial.rst]: @findepi @martint @nezihyigitbasi  I chose the name to *match* PostGIS which defines ST_IsValid and ST_IsValidReason: https://postgis.net/docs/ST_IsValidReason.html . ST_IsValidReason works well with ST_IsValid, but it is not a standard function. In the past Martin asked not to name non-standard functions in a way which suggests they are, e.g. ST_IsValidReason, hence, the `geometry_is_valid_reason` name. I'm happy to change it to whatever makes sense, but would like to hear what others think first. 

Nezih, Martin, do you have an opinion?

Comment (findepi) [reply presto-docs/src/main/sphinx/functions/geospatial.rst]: all i wanted to say is that `geometry_is_valid_reason` sounds like "returns reason why it is considered valid", while it "returns reason why it is considered *not* valid"

Comment (mbasmanova) [reply presto-docs/src/main/sphinx/functions/geospatial.rst]: @findepi Updated function name to `geometry_invalid_reason`

File: presto-geospatial-toolkit/src/main/java/com/facebook/presto/geospatial/GeometryJtsUtils.java [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+        }
+
+        for (int i = 0; i < partCount; i++) {
+            lineStrings[i] = FACTORY.createLineString(readCoordinates(input, partLengths[i]));
+        }
+
+        if (lineStrings.length == 1) {
+            return lineStrings[0];
+        }
+        else {
```
Comment (nezihyigitbasi) [inline presto-geospatial-toolkit/src/main/java/com/facebook/presto/geospatial/GeometryJtsUtils.java]: `else` not needed.

File: presto-geospatial-toolkit/src/main/java/com/facebook/presto/geospatial/GeometryJtsUtils.java [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+        for (int i = 0; i < pointCount; i++) {
+            points[i] = readPoint(input);
+        }
+        return FACTORY.createMultiPoint(points);
+    }
+
+    private static Geometry readPolyline(SliceInput input)
+    {
+        skipEnvelope(input);
+        int partCount = input.readInt();
```
Comment (nezihyigitbasi) [inline presto-geospatial-toolkit/src/main/java/com/facebook/presto/geospatial/GeometryJtsUtils.java]: You don't need to read `pointCount` if `partCount` is 0 (applies to the `readPolygon` method below)

```
int partCount = input.readInt();
if (partCount == 0) {
    return FACTORY.createPolygon();
}
```

File: presto-geospatial-toolkit/src/main/java/com/facebook/presto/geospatial/GeometryJtsUtils.java [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+        }
+        else if (geometries.size() == 1) {
+            return geometries.get(0);
+        }
+        return FACTORY.createGeometryCollection(geometries.toArray(new Geometry[0]));
+    }
+
+    private static Geometry readGeometry(SliceInput input)
+    {
+        int geometryType = input.readInt();
```
Comment (nezihyigitbasi) [inline presto-geospatial-toolkit/src/main/java/com/facebook/presto/geospatial/GeometryJtsUtils.java]: `requireNonNull(input, "input is null")` (applies to other methods in this class)

File: presto-geospatial-toolkit/src/main/java/com/facebook/presto/geospatial/GeometryJtsUtils.java [hunk 5]
```diff
@@ -0,0 +1,228 @@
...
+import static io.airlift.slice.SizeOf.SIZE_OF_DOUBLE;
+import static io.airlift.slice.SizeOf.SIZE_OF_INT;
+
+public class GeometryJtsUtils
+{
+    private static final int POINT = 1;
+    private static final int POLYLINE = 3;
+    private static final int POLYGON = 5;
+    private static final int MULTIPOINT = 8;
+    private static final GeometryFactory FACTORY = new GeometryFactory();
```
Comment (nezihyigitbasi) [inline presto-geospatial-toolkit/src/main/java/com/facebook/presto/geospatial/GeometryJtsUtils.java]: `FACTORY` -> `GEOMETRY_FACTORY`

File: presto-geospatial-toolkit/src/main/java/com/facebook/presto/geospatial/GeometryJtsUtils.java [hunk 6]
```diff
@@ -0,0 +1,228 @@
...
+import java.util.ArrayList;
+import java.util.List;
+
+import static com.facebook.presto.geospatial.GeometryUtils.isEsriNaN;
+import static io.airlift.slice.SizeOf.SIZE_OF_DOUBLE;
+import static io.airlift.slice.SizeOf.SIZE_OF_INT;
+
+public class GeometryJtsUtils
+{
+    private static final int POINT = 1;
```
Comment (nezihyigitbasi) [inline presto-geospatial-toolkit/src/main/java/com/facebook/presto/geospatial/GeometryJtsUtils.java]: We can create an enum for these four types.

File: presto-geospatial-toolkit/src/main/java/com/facebook/presto/geospatial/GeometryJtsUtils.java [hunk 7]
```diff
@@ -0,0 +1,228 @@
...
+import org.locationtech.jts.geom.Polygon;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import static com.facebook.presto.geospatial.GeometryUtils.isEsriNaN;
+import static io.airlift.slice.SizeOf.SIZE_OF_DOUBLE;
+import static io.airlift.slice.SizeOf.SIZE_OF_INT;
+
+public class GeometryJtsUtils
```
Comment (nezihyigitbasi) [inline presto-geospatial-toolkit/src/main/java/com/facebook/presto/geospatial/GeometryJtsUtils.java]: `JtsGeometryUtils` sounds more natural to me.

File: presto-geospatial-toolkit/src/main/java/com/facebook/presto/geospatial/GeometryJtsUtils.java [hunk 8]
```diff
@@ -0,0 +1,228 @@
...
+        input.skip(SIZE_OF_INT); // SRID
+
+        // GeometryCollection: spatialReferenceId|len-of-shape1|bytes-of-shape1|len-of-shape2|bytes-of-shape2...
+        List<Geometry> geometries = new ArrayList<>();
+        while (input.available() > 0) {
+            int length = input.readInt();
+            geometries.add(readGeometry(input.readSlice(length).getInput()));
+        }
+
+        if (geometries.isEmpty()) {
```
Comment (nezihyigitbasi) [inline presto-geospatial-toolkit/src/main/java/com/facebook/presto/geospatial/GeometryJtsUtils.java]: We can check `input.available() <= 0` before the `while` loop above and return `FACTORY.createGeometryCollection()`, that will simplify this `if/else` logic here.

File: presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/TestGeometryJtsUtils.java [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+            throws Exception
+    {
+        testGeometry("GEOMETRYCOLLECTION EMPTY");
+        testGeometry("GEOMETRYCOLLECTION (POINT (1 2), LINESTRING (0 0, 1 2, 3 4), POLYGON ((0 0, 0 1, 1 1, 1 0, 0 0)))");
+    }
+
+    private static void testGeometry(String wkt)
+            throws ParseException
+    {
+        Slice slice = stGeometryFromText(Slices.utf8Slice(wkt));
```
Comment (nezihyigitbasi) [inline presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/TestGeometryJtsUtils.java]: static import `utf8Slice`

Comment (nezihyigitbasi) [inline presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/TestGeometryJtsUtils.java]: rename `slice` -> `geometry`

File: presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/TestGeometryJtsUtils.java [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+        Geometry expected = WKT_READER.read(wkt);
+        Geometry actual = GeometryJtsUtils.deserialize(slice);
+
+        // ESRI shape serialization format doesn't contain enough information
+        // to distinguish between empty LineString and MultiLineString or
+        // empty Polygon and MultiPolygon.
+        if (expected.isEmpty()) {
```
Comment (nezihyigitbasi) [inline presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/TestGeometryJtsUtils.java]: If expected is empty you check whether actual is empty too, but doesn't the `assertEquals(actual, expected)` below handle empty geometries?

File: presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/GeoFunctions.java [hunk 11]
```diff
@@ -228,6 +231,32 @@ public static Boolean stIsEmpty(@SqlType(GEOMETRY_TYPE_NAME) Slice input)
...
 
+    @Description("Returns true if the input geometry is well formed")
+    @ScalarFunction("ST_IsValid")
+    @SqlType(StandardTypes.BOOLEAN)
+    public static boolean stIsValid(@SqlType(GEOMETRY_TYPE_NAME) Slice input)
+    {
+        return GeometryJtsUtils.deserialize(input).isValid();
+    }
+
+    @Description("If input geometry is not valid, return a reason why. Otherwise, returns null.")
```
Comment (nezihyigitbasi) [inline presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/GeoFunctions.java]: To make the description consistent with the other functions, how about `Returns the reason for why the input geometry is not valid. Returns null if the input is valid`?

File: presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/GeoFunctions.java [hunk 12]
```diff
@@ -228,6 +231,32 @@ public static Boolean stIsEmpty(@SqlType(GEOMETRY_TYPE_NAME) Slice input)
...
+    @ScalarFunction("geometry_is_valid_reason")
+    @SqlType(StandardTypes.VARCHAR)
+    @SqlNullable
+    public static Slice isValidReason(@SqlType(GEOMETRY_TYPE_NAME) Slice input)
```
Comment (nezihyigitbasi) [inline presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/GeoFunctions.java]: `isValidReason` -> `getInvalidReason`

File: presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/GeoFunctions.java [hunk 13]
```diff
@@ -228,6 +231,32 @@ public static Boolean stIsEmpty(@SqlType(GEOMETRY_TYPE_NAME) Slice input)
...
+    @Description("Returns true if the input geometry is well formed")
+    @ScalarFunction("ST_IsValid")
+    @SqlType(StandardTypes.BOOLEAN)
+    public static boolean stIsValid(@SqlType(GEOMETRY_TYPE_NAME) Slice input)
+    {
+        return GeometryJtsUtils.deserialize(input).isValid();
+    }
+
+    @Description("If input geometry is not valid, return a reason why. Otherwise, returns null.")
+    @ScalarFunction("geometry_is_valid_reason")
```
Comment (nezihyigitbasi) [inline presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/GeoFunctions.java]: This function returns a varchar/reason string, but the name `geometry_is_valid_reason` sounds like it returns a boolean (`is...`).

File: presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/GeoFunctions.java [hunk 14]
```diff
@@ -228,6 +231,32 @@ public static Boolean stIsEmpty(@SqlType(GEOMETRY_TYPE_NAME) Slice input)
...
+    @SqlType(StandardTypes.VARCHAR)
+    @SqlNullable
+    public static Slice isValidReason(@SqlType(GEOMETRY_TYPE_NAME) Slice input)
+    {
+        if (input == null) {
+            return null;
+        }
+        Geometry geometry = GeometryJtsUtils.deserialize(input);
```
Comment (nezihyigitbasi) [inline presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/GeoFunctions.java]: `geometry` can be null (deserialize can return null).

File: presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/GeoFunctions.java [hunk 15]
```diff
@@ -228,6 +231,32 @@ public static Boolean stIsEmpty(@SqlType(GEOMETRY_TYPE_NAME) Slice input)
...
+        IsValidOp isValidOp = new IsValidOp(geometry);
```
Comment (nezihyigitbasi) [inline presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/GeoFunctions.java]: `isValid` does a validation under the hood with `checkValid`. Then we do another validation with `getValidationError` below. We can get away with a single validation:
```
    TopologyValidationError error = new IsValidOp(geometry).getValidationError(); // got rid of the isValidOp local variable
    if (error == null) {
        return null;
    }
    return utf8Slice(error.toString()); // added static import
```

File: presto-docs/src/main/sphinx/functions/geospatial.rst [hunk 16]
```diff
@@ -159,6 +159,12 @@ Accessors
 
     Returns ``true`` if and only if the line is closed and simple.
 
+.. function:: ST_IsValid(Geometry) -> boolean
+
+    Returns ``true`` if and only if the input geometry is well formed.
+    geometry_invalid_reason function returns the reason why geometry is
```
Comment (nezihyigitbasi) [inline presto-docs/src/main/sphinx/functions/geospatial.rst]: ```` ``geometry_invalid_reason`` ````

File: presto-docs/src/main/sphinx/functions/geospatial.rst [hunk 17]
```diff
@@ -159,6 +159,12 @@ Accessors
...
+    not well formed or null if it is.
```
Comment (nezihyigitbasi) [inline presto-docs/src/main/sphinx/functions/geospatial.rst]: ```` ``NULL`` ````

File: presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/TestGeoFunctions.java [hunk 18]
```diff
@@ -164,6 +164,64 @@ public void testSTIsEmpty()
...
+        // invalid geometries
+        assertFunction("ST_IsValid(ST_GeometryFromText('POLYGON ((0 0, 0 1, 1 1, 1 0, 0 0), (2 2, 2 3, 3 3, 3 2, 2 2))'))", BOOLEAN, false);
+        assertFunction("ST_IsValid(ST_GeometryFromText('POLYGON ((0 0, 0 1, 2 1, 1 1, 1 0, 0 0))'))", BOOLEAN, false);
+        assertFunction("ST_IsValid(ST_GeometryFromText('POLYGON ((0 0, 0 1, 1 1, 1 0, 0 0), (0 1, 1 1, 0.5 0.5, 0 1))'))", BOOLEAN, false);
+        assertFunction("ST_IsValid(ST_GeometryFromText('MULTIPOLYGON (((0 0, 0 1, 1 1, 1 0, 0 0)), ((0.5 0.5, 0.5 2, 2 2, 2 0.5, 0.5 0.5)))'))", BOOLEAN, false);
+        assertFunction("ST_IsValid(ST_GeometryFromText('GEOMETRYCOLLECTION (POINT (1 2), POLYGON ((0 0, 0 1, 2 1, 1 1, 1 0, 0 0)))'))", BOOLEAN, false);
+    }
+
+    @Test
+    public void testInvalidReason()
```
Comment (nezihyigitbasi) [inline presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/TestGeoFunctions.java]: We can shorten these two tests. For the valid geometries you can add a helper method (`assertValidGeometry`) to assert that they are valid AND their invalid reasons are null, so you can assert that the reason is null in `testSTIsValid` above in one go. And in `testInvalidReason` you can just test invalid geometries.

File: presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/GeoFunctions.java [hunk 19]
```diff
@@ -228,6 +231,32 @@ public static Boolean stIsEmpty(@SqlType(GEOMETRY_TYPE_NAME) Slice input)
         return deserialize(input).isEmpty();
     }
 
+    @Description("Returns true if the input geometry is well formed")
+    @ScalarFunction("ST_IsValid")
+    @SqlType(StandardTypes.BOOLEAN)
+    public static boolean stIsValid(@SqlType(GEOMETRY_TYPE_NAME) Slice input)
+    {
+        return GeometryJtsUtils.deserialize(input).isValid();
```
Comment (nezihyigitbasi) [inline presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/GeoFunctions.java]: `deserialize` can return null.

File: presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/TestGeometryJtsUtils.java (hunk 10 above)
Comment (mbasmanova) [reply presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/TestGeometryJtsUtils.java]: `assertEquals(actual, expected)` fails if both geometries are empty, but one is LineString and the other is a MultiLineString.

File: presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/GeoFunctions.java (hunk 14 above)
Comment (mbasmanova) [reply presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/GeoFunctions.java]: `deserialize` returns null only for null input, but I'll move the check for null after call to deserialize for clarity.

**Comment by nezihyigitbasi** [comment]
`presto-raptor` test failure is due to #6434.

**Comment by mbasmanova** [comment]
@nezihyigitbasi Nezih, thank you for review. I addressed your comments and updated this PR.

File: presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/TestJtsGeometryUtils.java [hunk 20]
```diff
@@ -0,0 +1,107 @@
...
+
+    @Test
+    public void testGeometryCollection()
+            throws Exception
+    {
+        testGeometry("GEOMETRYCOLLECTION EMPTY");
+        testGeometry("GEOMETRYCOLLECTION (POINT (1 2), LINESTRING (0 0, 1 2, 3 4), POLYGON ((0 0, 0 1, 1 1, 1 0, 0 0)))");
+    }
+
+    private static void testGeometry(String wkt)
```
Comment (nezihyigitbasi) [inline presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/TestJtsGeometryUtils.java]: `testGeometry` -> `assertGeometry`

File: presto-geospatial-toolkit/src/main/java/com/facebook/presto/geospatial/JtsGeometryUtils.java [hunk 21]
```diff
@@ -0,0 +1,262 @@
...
+
+import static com.facebook.presto.geospatial.GeometryUtils.isEsriNaN;
+import static io.airlift.slice.SizeOf.SIZE_OF_DOUBLE;
+import static io.airlift.slice.SizeOf.SIZE_OF_INT;
+import static java.util.Objects.requireNonNull;
+
+public class JtsGeometryUtils
+{
+    private enum EsriShapeTypes {
+        POINT(1),
```
[the remark this answers is not in the document]
Comment (mbasmanova) [reply presto-geospatial-toolkit/src/main/java/com/facebook/presto/geospatial/JtsGeometryUtils.java]: These numbers come from ESRI's specification : https://www.esri.com/library/whitepapers/pdfs/shapefile.pdf

"Shape types not specified above (2, 4, 6, etc., and up to 33) are reserved for future use."

File: presto-geospatial-toolkit/src/main/java/com/facebook/presto/geospatial/JtsGeometryUtils.java [hunk 22]
```diff
@@ -0,0 +1,262 @@
...
+        }
+
+        int pointCount = input.readInt();
+        int[] startIndexes = new int[partCount];
+        for (int i = 0; i < partCount; i++) {
+            startIndexes[i] = input.readInt();
+        }
+
+        int[] partLengths = new int[partCount];
+        if (partCount > 1) {
```
[the remark this answers is not in the document]
Comment (mbasmanova) [reply presto-geospatial-toolkit/src/main/java/com/facebook/presto/geospatial/JtsGeometryUtils.java]: Looks correct to me. This is the most common case and it is covered by tests.

File: presto-geospatial-toolkit/src/main/java/com/facebook/presto/geospatial/JtsGeometryUtils.java (hunk 21 above)
[the remark this answers is not in the document]
Comment (nezihyigitbasi) [reply presto-geospatial-toolkit/src/main/java/com/facebook/presto/geospatial/JtsGeometryUtils.java]: Please add a comment saying these codes match the ESRI spec, otherwise they seem arbitrary at first look.

**Comment by mbasmanova** [comment]
@nezihyigitbasi Nezih, thank you for the review. Updated the PR accordingly.

File: presto-docs/src/main/sphinx/functions/geospatial.rst [hunk 23]
```diff
@@ -163,6 +163,12 @@ Accessors
 
     Returns ``true`` if and only if the line is closed and simple.
 
+.. function:: ST_IsValid(Geometry) -> boolean
+
+    Returns ``true`` if and only if the input geometry is well formed.
+    ``geometry_invalid_reason`` function returns the reason why geometry is
```
Comment (electrum) [inline presto-docs/src/main/sphinx/functions/geospatial.rst]: This should be documented as a separate function. Make this a function link.

Comment (electrum) [reply presto-docs/src/main/sphinx/functions/geospatial.rst]: Let's rephrase this as
```
Use :func:`geometry_invalid_reason` to determine why the geometry is not well formed.

**Comment by mbasmanova** [comment]
@electrum David, thanks for the suggestion. I updated documentation.

**Comment by mbasmanova** [comment]
@electrum David, I made the changes you requested. Is there anything else that needs updating? If not, would you approve this PR?

File: presto-geospatial-toolkit/src/main/java/com/facebook/presto/geospatial/JtsGeometryUtils.java [hunk 24]
```diff
@@ -0,0 +1,260 @@
...
+        }
+        return GEOMETRY_FACTORY.createMultiLineString(lineStrings);
+    }
+
+    private static Coordinate[] readCoordinates(SliceInput input, int count)
+    {
+        requireNonNull(input, "input is null");
+        verify(count > 0);
+        Coordinate[] coordinates = new Coordinate[count];
+        for (int j = 0; j < count; j++) {
```
Comment (electrum) [inline presto-geospatial-toolkit/src/main/java/com/facebook/presto/geospatial/JtsGeometryUtils.java]: Name this `i`
